### PR TITLE
Fix Binance too many connection attempts problem.

### PR DESCRIPTION
**Pull request by emreeroglu**
## Description
Time to time Binance does not respond API calls and this is crashing the bot.
```
AttributeError("'PublicAPI' object has no attribute 'client'")
Traceback (most recent call last):
  File "/root/bot/USDTTRY/pycryptobot.py", line 1073, in <module>
    main()
  File "/root/bot/USDTTRY/pycryptobot.py", line 1036, in main
    runApp()
  File "/root/bot/USDTTRY/pycryptobot.py", line 1033, in runApp
    s.run()
  File "/usr/lib64/python3.9/sched.py", line 151, in run
    action(*argument, **kwargs)
  File "/root/bot/USDTTRY/pycryptobot.py", line 201, in executeJob
    trading_data = app.getHistoricalData(app.getMarket(), app.getGranularity())
  File "/root/bot/USDTTRY/models/PyCryptoBot.py", line 236, in getHistoricalData
    return api.getHistoricalData(market, granularity)
  File "/root/bot/USDTTRY/models/exchange/binance/api.py", line 381, in getHistoricalData
    resp = self.client.get_historical_klines(market, granularity, '13 days ago UTC')
AttributeError: 'PublicAPI' object has no attribute 'client'
```
Fixes # (issue)
Added a for loop to places I had the problem which is trying 10 times to create client and then if its still not successful raise an Error.

## Type of change

Please make your selection.

- [x ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] This change requires a new dependency
- [ ] Code Maintainability / comments
- [ ] Code Refactoring / future-proofing

## How Has This Been Tested?
Made live trading and watched logs, instead of crashing app managed get client at first try most of the time, and continue working normally.

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

## Checklist:

- [ x ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added pytest unit tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: models/exchange/binance/api.py [hunk 1]
```diff
@@ -65,7 +67,17 @@ def __init__(self, api_key: str='', api_secret: str='', api_url: str='https://ap
...
+        for i in range(10):
+            try:
+                self.client = Client(self.api_key, self.api_secret, { 'verify': False, 'timeout': 20 })
+                break
+            except:
+                if i == 9:
+                    raise SystemExit("Can not create instance of AuthAPI client.")
+                print('Exception: ', sys.exc_info()[0]) 
+                print('Error on creating instance of AuthAPI Client. Trying again... Attempt: ', i)
+                sleep(0.1)
```
Comment (whittlem) [inline models/exchange/binance/api.py:80]: 0.1 is very short. Would it not be better to set say 5-10 seconds just to give it a chance to respond?

File: models/exchange/binance/api.py [hunk 2]
```diff
@@ -65,7 +67,17 @@ def __init__(self, api_key: str='', api_secret: str='', api_url: str='https://ap
         self.api_url = api_url
         self.api_key = api_key
         self.api_secret = api_secret
-        self.client = Client(self.api_key, self.api_secret, { 'verify': False, 'timeout': 20 })
```
Comment (whittlem) [inline models/exchange/binance/api.py:68]: Also wondering if there are latency issues with Binance to maybe increase the "timeout" to maybe 60. I mean 20 seconds is long! But if the exchange is struggling maybe giving it a bit more time is better.

**Comment by emreeroglu** [comment]
```

ADAEUR.log ConnectionError(MaxRetryError("HTTPSConnectionPool(host='api.binance.com', port=443): Max retries exceeded with url: /api/v3/ping (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f407c9ec160>: Failed to establish a new connection: [Errno -2] Name or service not known'))"))
ADAEUR.log Traceback (most recent call last):
ADAEUR.log   File "/root/bot/.venv/lib64/python3.9/site-packages/urllib3/connection.py", line 159, in _new_conn
ADAEUR.log     conn = connection.create_connection(
ADAEUR.log   File "/root/bot/.venv/lib64/python3.9/site-packages/urllib3/util/connection.py", line 61, in create_connection
ADAEUR.log     for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
ADAEUR.log   File "/usr/lib64/python3.9/socket.py", line 953, in getaddrinfo
ADAEUR.log     for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
ADAEUR.log socket.gaierror: [Errno -2] Name or service not known
ADAEUR.log 
ADAEUR.log During handling of the above exception, another exception occurred:
ADAEUR.log 
ADAEUR.log Traceback (most recent call last):
ADAEUR.log   File "/root/bot/.venv/lib64/python3.9/site-packages/urllib3/connectionpool.py", line 670, in urlopen
ADAEUR.log     httplib_response = self._make_request(
ADAEUR.log   File "/root/bot/.venv/lib64/python3.9/site-packages/urllib3/connectionpool.py", line 381, in _make_request
ADAEUR.log     self._validate_conn(conn)
ADAEUR.log   File "/root/bot/.venv/lib64/python3.9/site-packages/urllib3/connectionpool.py", line 978, in _validate_conn
ADAEUR.log     conn.connect()
ADAEUR.log   File "/root/bot/.venv/lib64/python3.9/site-packages/urllib3/connection.py", line 309, in connect
ADAEUR.log     conn = self._new_conn()
ADAEUR.log   File "/root/bot/.venv/lib64/python3.9/site-packages/urllib3/connection.py", line 171, in _new_conn
ADAEUR.log     raise NewConnectionError(
ADAEUR.log urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPSConnection object at 0x7f407c9ec160>: Failed to establish a new connection: [Errno -2] Name or service not known
ADAEUR.log 
ADAEUR.log During handling of the above exception, another exception occurred:
ADAEUR.log 
ADAEUR.log Traceback (most recent call last):
ADAEUR.log   File "/root/bot/.venv/lib64/python3.9/site-packages/requests/adapters.py", line 439, in send
ADAEUR.log     resp = conn.urlopen(
ADAEUR.log   File "/root/bot/.venv/lib64/python3.9/site-packages/urllib3/connectionpool.py", line 726, in urlopen
ADAEUR.log     retries = retries.increment(
ADAEUR.log   File "/root/bot/.venv/lib64/python3.9/site-packages/urllib3/util/retry.py", line 446, in increment
ADAEUR.log     raise MaxRetryError(_pool, url, error or ResponseError(cause))
ADAEUR.log urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='api.binance.com', port=443): Max retries exceeded with url: /api/v3/ping (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f407c9ec160>: Failed to establish a new connection: [Errno -2] Name or service not known'))
ADAEUR.log 
ADAEUR.log During handling of the above exception, another exception occurred:
ADAEUR.log 
ADAEUR.log Traceback (most recent call last):
ADAEUR.log   File "/root/bot/ADAEUR/pycryptobot.py", line 1083, in <module>
ADAEUR.log     main()
ADAEUR.log   File "/root/bot/ADAEUR/pycryptobot.py", line 1045, in main
ADAEUR.log     runApp()
ADAEUR.log   File "/root/bot/ADAEUR/pycryptobot.py", line 1042, in runApp
ADAEUR.log     s.run()
ADAEUR.log   File "/usr/lib64/python3.9/sched.py", line 151, in run
ADAEUR.log     action(*argument, **kwargs)
ADAEUR.log   File "/root/bot/ADAEUR/pycryptobot.py", line 274, in executeJob
ADAEUR.log     ticker = app.getTicker(app.getMarket())
ADAEUR.log   File "/root/bot/ADAEUR/models/PyCryptoBot.py", line 402, in getTicker
ADAEUR.log     api = BPublicAPI()
ADAEUR.log   File "/root/bot/ADAEUR/models/exchange/binance/api.py", line 322, in __init__
ADAEUR.log     self.client = Client()
ADAEUR.log   File "/root/bot/.venv/lib64/python3.9/site-packages/binance/client.py", line 105, in __init__
ADAEUR.log     self.ping()
ADAEUR.log   File "/root/bot/.venv/lib64/python3.9/site-packages/binance/client.py", line 392, in ping
ADAEUR.log     return self._get('ping', version=self.PRIVATE_API_VERSION)
ADAEUR.log   File "/root/bot/.venv/lib64/python3.9/site-packages/binance/client.py", line 237, in _get
ADAEUR.log     return self._request_api('get', path, signed, version, **kwargs)
ADAEUR.log   File "/root/bot/.venv/lib64/python3.9/site-packages/binance/client.py", line 202, in _request_api
ADAEUR.log     return self._request(method, uri, signed, **kwargs)
ADAEUR.log   File "/root/bot/.venv/lib64/python3.9/site-packages/binance/client.py", line 196, in _request
ADAEUR.log     self.response = getattr(self.session, method)(uri, **kwargs)
ADAEUR.log   File "/root/bot/.venv/lib64/python3.9/site-packages/requests/sessions.py", line 555, in get
ADAEUR.log     return self.request('GET', url, **kwargs)
ADAEUR.log   File "/root/bot/.venv/lib64/python3.9/site-packages/requests/sessions.py", line 542, in request
ADAEUR.log     resp = self.send(prep, **send_kwargs)
ADAEUR.log   File "/root/bot/.venv/lib64/python3.9/site-packages/requests/sessions.py", line 655, in send
ADAEUR.log     r = adapter.send(request, **kwargs)
ADAEUR.log   File "/root/bot/.venv/lib64/python3.9/site-packages/requests/adapters.py", line 516, in send
ADAEUR.log     raise ConnectionError(e, request=request)
ADAEUR.log requests.exceptions.ConnectionError: HTTPSConnectionPool(host='api.binance.com', port=443): Max retries exceeded with url: /api/v3/ping (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f407c9ec160>: Failed to establish a new connection: [Errno -2] Name or service not known'))

```

**Comment by emreeroglu** [comment]
```
XLMBUSD got an exception: ConnectionError(MaxRetryError("HTTPSConnectionPool(host='api.binance.com', port=443): Max retries exceeded with url: /api/v3/ping (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f5074bd0bb0>: Failed to establish a new connection: [Errno -2] Name or service not known'))"))
XLMBUSD.log ConnectionError(MaxRetryError("HTTPSConnectionPool(host='api.binance.com', port=443): Max retries exceeded with url: /api/v3/ping (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f5074bd0bb0>: Failed to establish a new connection: [Errno -2] Name or service not known'))"))
XLMBUSD.log Traceback (most recent call last):
XLMBUSD.log   File "/root/bot/.venv/lib64/python3.9/site-packages/urllib3/connection.py", line 159, in _new_conn
XLMBUSD.log     conn = connection.create_connection(
XLMBUSD.log   File "/root/bot/.venv/lib64/python3.9/site-packages/urllib3/util/connection.py", line 61, in create_connection
XLMBUSD.log     for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
XLMBUSD.log   File "/usr/lib64/python3.9/socket.py", line 953, in getaddrinfo
XLMBUSD.log     for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
XLMBUSD.log socket.gaierror: [Errno -2] Name or service not known
XLMBUSD.log 
XLMBUSD.log During handling of the above exception, another exception occurred:
XLMBUSD.log 
XLMBUSD.log Traceback (most recent call last):
XLMBUSD.log   File "/root/bot/.venv/lib64/python3.9/site-packages/urllib3/connectionpool.py", line 670, in urlopen
XLMBUSD.log     httplib_response = self._make_request(
XLMBUSD.log   File "/root/bot/.venv/lib64/python3.9/site-packages/urllib3/connectionpool.py", line 381, in _make_request
XLMBUSD.log     self._validate_conn(conn)
XLMBUSD.log   File "/root/bot/.venv/lib64/python3.9/site-packages/urllib3/connectionpool.py", line 978, in _validate_conn
XLMBUSD.log     conn.connect()
XLMBUSD.log   File "/root/bot/.venv/lib64/python3.9/site-packages/urllib3/connection.py", line 309, in connect
XLMBUSD.log     conn = self._new_conn()
XLMBUSD.log   File "/root/bot/.venv/lib64/python3.9/site-packages/urllib3/connection.py", line 171, in _new_conn
XLMBUSD.log     raise NewConnectionError(
XLMBUSD.log urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPSConnection object at 0x7f5074bd0bb0>: Failed to establish a new connection: [Errno -2] Name or service not known
XLMBUSD.log 
XLMBUSD.log During handling of the above exception, another exception occurred:
XLMBUSD.log 
XLMBUSD.log Traceback (most recent call last):
XLMBUSD.log   File "/root/bot/.venv/lib64/python3.9/site-packages/requests/adapters.py", line 439, in send
XLMBUSD.log     resp = conn.urlopen(
XLMBUSD.log   File "/root/bot/.venv/lib64/python3.9/site-packages/urllib3/connectionpool.py", line 726, in urlopen
XLMBUSD.log     retries = retries.increment(
XLMBUSD.log   File "/root/bot/.venv/lib64/python3.9/site-packages/urllib3/util/retry.py", line 446, in increment
XLMBUSD.log     raise MaxRetryError(_pool, url, error or ResponseError(cause))
XLMBUSD.log urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='api.binance.com', port=443): Max retries exceeded with url: /api/v3/ping (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f5074bd0bb0>: Failed to establish a new connection: [Errno -2] Name or service not known'))
XLMBUSD.log 
XLMBUSD.log During handling of the above exception, another exception occurred:
XLMBUSD.log 
XLMBUSD.log Traceback (most recent call last):
XLMBUSD.log   File "/root/bot/XLMBUSD/pycryptobot.py", line 1031, in <module>
XLMBUSD.log     main()
XLMBUSD.log   File "/root/bot/XLMBUSD/pycryptobot.py", line 996, in main
XLMBUSD.log     runApp()
XLMBUSD.log   File "/root/bot/XLMBUSD/pycryptobot.py", line 993, in runApp
XLMBUSD.log     s.run()
XLMBUSD.log   File "/usr/lib64/python3.9/sched.py", line 151, in run
XLMBUSD.log     action(*argument, **kwargs)
XLMBUSD.log   File "/root/bot/XLMBUSD/pycryptobot.py", line 340, in executeJob
XLMBUSD.log     sell_taker_fee=app.getTakerFee(),
XLMBUSD.log   File "/root/bot/XLMBUSD/models/PyCryptoBot.py", line 540, in getTakerFee
XLMBUSD.log     api = BAuthAPI(self.getAPIKey(), self.getAPISecret(), self.getAPIURL())
XLMBUSD.log   File "/root/bot/XLMBUSD/models/exchange/binance/api.py", line 68, in __init__
XLMBUSD.log     self.client = Client(self.api_key, self.api_secret, { 'verify': False, 'timeout': 20 })
XLMBUSD.log   File "/root/bot/.venv/lib64/python3.9/site-packages/binance/client.py", line 105, in __init__
XLMBUSD.log     self.ping()
XLMBUSD.log   File "/root/bot/.venv/lib64/python3.9/site-packages/binance/client.py", line 392, in ping
XLMBUSD.log     return self._get('ping', version=self.PRIVATE_API_VERSION)
XLMBUSD.log   File "/root/bot/.venv/lib64/python3.9/site-packages/binance/client.py", line 237, in _get
XLMBUSD.log     return self._request_api('get', path, signed, version, **kwargs)
XLMBUSD.log   File "/root/bot/.venv/lib64/python3.9/site-packages/binance/client.py", line 202, in _request_api
XLMBUSD.log     return self._request(method, uri, signed, **kwargs)
XLMBUSD.log   File "/root/bot/.venv/lib64/python3.9/site-packages/binance/client.py", line 196, in _request
XLMBUSD.log     self.response = getattr(self.session, method)(uri, **kwargs)
XLMBUSD.log   File "/root/bot/.venv/lib64/python3.9/site-packages/requests/sessions.py", line 555, in get
XLMBUSD.log     return self.request('GET', url, **kwargs)
XLMBUSD.log   File "/root/bot/.venv/lib64/python3.9/site-packages/requests/sessions.py", line 542, in request
XLMBUSD.log     resp = self.send(prep, **send_kwargs)
XLMBUSD.log   File "/root/bot/.venv/lib64/python3.9/site-packages/requests/sessions.py", line 655, in send
XLMBUSD.log     r = adapter.send(request, **kwargs)
XLMBUSD.log   File "/root/bot/.venv/lib64/python3.9/site-packages/requests/adapters.py", line 516, in send
XLMBUSD.log     raise ConnectionError(e, request=request)
XLMBUSD.log requests.exceptions.ConnectionError: HTTPSConnectionPool(host='api.binance.com', port=443): Max retries exceeded with url: /api/v3/ping (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f5074bd0bb0>: Failed to establish a new connection: [Errno -2] Name or service not known'))
```

File: models/exchange/binance/api.py (hunk 1 above)
Comment (emreeroglu) [reply models/exchange/binance/api.py:80]: The logs below are shows that problem is not about to first connection to Exchange.
Its about to API request limit of the Binance.

If we put 5-10 second waiting there we will be stopping bot for that long.
I implemented and tested this and what I see if we catch exception;
most of the time we get the Client() object instance at first try.

**Review by whittlem — APPROVED** [review]
Ok, thanks for the update.

File: models/exchange/binance/api.py (hunk 1 above)
Comment (iThom) [reply models/exchange/binance/api.py:80]: Also the code is duplicated in `AuthAPI` and `PublicAPI` L329-340, to make maintenance easier, why to not use the inherited class `AuthAPIBase` just adding a method to get the client instance and after use it in both classes that already inherit this class ?

Comment (emreeroglu) [reply models/exchange/binance/api.py:80]: You are correct, this just a little bit saving the day,
We should not need to connecting APIs for every another call.
Auth can be handled in Public API constructor and PublicAPI can be a global variable, or part of something like state.
So we would not create an API conncetion for every request.

**Pull request merged**